nexus/src/app/update/mgs_clients.rs

```diff
@@ -0,0 +1,244 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct MgsClients {
+    clients: VecDeque<Arc<Client>>,
+}
+
+impl MgsClients {
+    /// Create a new `MgsClients` with the given `clients`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `clients` is empty.
+    pub fn new<T: Into<VecDeque<Arc<Client>>>>(clients: T) -> Self {
+        let clients = clients.into();
+        assert!(!clients.is_empty());
+        Self { clients }
+    }
+
+    /// Create a new `MgsClients` with the given `clients`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `clients` is empty.
+    pub fn from_clients<I: IntoIterator<Item = Client>>(clients: I) -> Self {
+        let clients = clients
+            .into_iter()
+            .map(Arc::new)
+            .collect::<VecDeque<Arc<Client>>>();
+        Self::new(clients)
+    }
+
+    /// Run `op` against all clients, returning either the first success, the
+    /// first non-communication error, or the communication error (if all
+    /// clients failed with communication errors).
+    ///
+    /// On a successful return, the internal client list will be reordered so
+    /// any future accesses will attempt the most-recently-successful client.
+    pub(super) async fn try_all<T, F, Fut>(
+        &mut self,
+        log: &Logger,
+        op: F,
+    ) -> Result<T, GatewayClientError>
+    where
+        // Seems like it would be nicer to take `&Client` here instead of
+        // needing to clone each `Arc`, but there's currently no decent way of
+        // doing that without boxing the returned future:
+        // https://users.rust-lang.org/t/how-to-express-that-the-future-returned-by-a-closure-lives-only-as-long-as-its-argument/90039/10
+        F: Fn(Arc<Client>) -> Fut,
+        Fut: Future<Output = Result<T, GatewayClientError>>,
+    {
+        let mut last_err = None;
+        for (i, client) in self.clients.iter().enumerate() {
+            match op(Arc::clone(client)).await {
+                Ok(value) => {
+                    self.clients.rotate_left(i);
+                    return Ok(value);
+                }
+                Err(GatewayClientError::CommunicationError(err)) => {
+                    if i < self.clients.len() {
+                        warn!(
+                            log, "communication error with MGS; \
+                                  will try next client";
+                            "mgs_addr" => client.baseurl(),
+                            "err" => %err,
+                        );
+                    }
+                    last_err = Some(err);
+                    continue;
+                }
+                Err(err) => return Err(err),
+            }
+        }
+
+        // The only way to get here is if all clients failed with communication
+        // errors. Return the error from the last MGS we tried.
+        Err(GatewayClientError::CommunicationError(last_err.unwrap()))
+    }
+
+    /// Poll for the status of an expected-to-be-in-progress update.
```

Comment:
It seems like this isn't really related to the multi-MGS thing.  I wonder if some of this functionality should go in gateway_client itself.

Reply:
Partially agreed - in a subsequent PR I end up moving this into a "common SP updating stuff" module. Maybe gateway-client is ultimately the right home for this, but that would also require moving `UpdateProgress` into some other crate that both gateway-client and nexus depend on (omicron-common maybe?). That seems fine to me but also easy to do / not urgent, and could happen at whatever point we start wanting to use `UpdateProgress` (which may need tweaking as we add other kinds of updates), I think?